### PR TITLE
feat: add createManifest API

### LINKED ISSUE #6624: add API to call createManifest endpoint
### Is your feature request related to a problem? Please describe

enhance libpod Dockerode with createManifest endpoint

https://docs.podman.io/en/latest/_static/api.html#tag/manifests/operation/ManifestCreateLibpod

### Describe the solution you'd like

this method will allow to create a manifest

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cdrage**
feat: add createManifest API

### What does this PR do?

Adds the createManifest API modelled by https://docs.podman.io/en/latest/_static/api.html#tag/manifests/operation/ManifestCreateLibpod

You'll be able to pass in createManifestOptions with the name as well as
an array of image names and be able to create a manifest.

### Screenshot / video of UI

N/A

### What issues does this PR fix or reference?

Closes https://github.com/containers/podman-desktop/issues/6624

### How to test this PR?

- [X] Tests are covering the bug fix or the new feature

Either through tests or you can also do the following in a svelte file:

```ts
    const options: ManifestCreateOptions = {
      name: containerImageName,
      images: [
        containerImageName],

    };
    await window.createManifest(options);
```

Signed-off-by: Charlie Drage <charlie@charliedrage.com>

### CODE REVIEW

**Comment by cdrage** [comment]
Could you do a preliminary review @benoitf ?

I have not written tests yet, but want to make sure this is a correct API implementation.

FYI it returns Promise<void> as it's modeled after the actual call: https://docs.podman.io/en/latest/_static/api.html#tag/manifests/operation/ManifestCreateLibpod similar to our other API calls.

Otherwise, this works and I can confirm it's working via:

```ts
    const options: ManifestCreateOptions = {
      name: containerImageName,
      images: [
        containerImageName],

    };
    await window.createManifest(options);
```

File: packages/main/src/plugin/container-registry.ts [hunk 1]
```diff
@@ -1224,6 +1224,29 @@ export class ContainerProviderRegistry {
     }
   }
 
+  async createManifest(manifestOptions: containerDesktopAPI.ManifestCreateOptions): Promise<void> {
```
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts]: it should not be ManifestCreateOptions from the desktopAPI there but from Manifest-info.ts

File: packages/main/src/plugin/container-registry.ts [hunk 2]
```diff
@@ -1224,6 +1224,29 @@ export class ContainerProviderRegistry {
...
+      if (manifestOptions.provider) {
+        // grab connection
+        internalContainerProvider = this.getMatchingContainerProvider(manifestOptions.provider);
+      } else {
+        // Get the first running podman connection
+        internalContainerProvider = this.getFirstRunningPodmanContainerProvider();
+      }
+      if (!internalContainerProvider?.libpodApi) {
+        throw new Error('No podman provider with a running engine');
+      }
```
Comment (benoitf) [inline packages/main/src/plugin/container-registry.ts]: if it's finding a provider but the provider is not supporting the podman API (like docker) I should not get the error message that no podman provider is running.


we need to differentiate: not finding a provider or not having the libpodAPI on the matching provider

File: packages/main/src/plugin/dockerode/libpod-dockerode.ts [hunk 3]
```diff
@@ -777,5 +779,29 @@ export class LibpodDockerode {
         });
       });
     };
+
+    // add createManifest
+    prototypeOfDockerode.createManifest = function (manifestOptions: ManifestCreateOptions): Promise<unknown> {
```
Comment (benoitf) [inline packages/main/src/plugin/dockerode/libpod-dockerode.ts]:
```suggestion
    prototypeOfDockerode.podmanCreateManifest = function (manifestOptions: ManifestCreateOptions): Promise<unknown> {
```

Comment (benoitf) [reply packages/main/src/plugin/dockerode/libpod-dockerode.ts]: need to prefix with podman

File: packages/main/src/plugin/index.ts [hunk 4]
```diff
@@ -750,6 +750,15 @@ export class PluginSystem {
         return containerProviderRegistry.removePod(engine, podId);
       },
     );
+
+    // manifest
+    this.ipcHandle(
+      'container-provider-registry:createManifest',
+      async (_listener, manifestOptions: containerDesktopAPI.ManifestCreateOptions): Promise<void> => {
```
Comment (benoitf) [inline packages/main/src/plugin/index.ts]: not the extension api there

File: packages/preload/src/index.ts [hunk 5]
```diff
@@ -282,6 +282,11 @@ export function initExposure(): void {
     return ipcInvoke('container-provider-registry:restartPod', engine, podId);
   });
 
+  // Manifest
+  contextBridge.exposeInMainWorld('createManifest', async (createOptions: containerDesktopAPI.ManifestCreateOptions): Promise<void> => {
```
Comment (benoitf) [inline packages/preload/src/index.ts]: not the extension api, should be from manifest info

**Comment by cdrage** [comment]
Updates:
- ~FYI getting some odd issues with the API endpoint, so hold off on testing via window.createManifest~ clarified with https://github.com/containers/podman/issues/22255

TODO:
- [x] Add undocumented callback...
```ts
▶ curl --unix-socket /var/folders/cn/pm5mgrcs5h339nphypxtvd080000gn/T/podman/podman-machine-default-api.sock -X POST \
     "http://d/v4.2.0/libpod/manifests/test?images=quay.io/bootc-extension/httpd"

{"Id":"6a105d8fe5a76e15c4687f44deb923acf62608d56ede79b74f5a9251de5e2155"}
```

Returns ID which is undocumented in the API.
- [x] Add tests

**Comment by cdrage** [comment]
Opened up issue https://github.com/containers/podman/issues/22255 for podman upstream for clarity on the API.

**Comment by cdrage** [comment]
Tests have been added and return value corrected, ready for review!

**Comment by cdrage** [comment]
Tests have been added and return value corrected, ready for review!

File: packages/extension-api/src/extension-api.d.ts [hunk 6]
```diff
@@ -285,6 +285,14 @@ declare module '@podman-desktop/api' {
     provider?: ContainerProviderConnection;
   }
 
+  export interface ManifestCreateOptions {
+    images: string[];
+    name: string;
+
```
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: we're missing some properties

`all` and `amend`

File: packages/extension-api/src/extension-api.d.ts [hunk 7]
```diff
@@ -3367,6 +3377,9 @@ declare module '@podman-desktop/api' {
     export function listPods(): Promise<PodInfo[]>;
     export function stopPod(engineId: string, podId: string): Promise<void>;
     export function removePod(engineId: string, podId: string): Promise<void>;
+
+    // Manifest related methods
+    export function createManifest(options: ManifestCreateOptions): Promise<{ Id: string }>;
```
Comment (benoitf) [inline packages/extension-api/src/extension-api.d.ts]: it might help if we return also the engineId

**Review by benoitf — APPROVED** [review]
LGTM (return the engineId in addition to the Id, because you may need the engineId when you may want to inspect it later)

If you could use fixup commits rather than always pushing a single commit during the review, it would help to know the diff.

Now, each time I need to review the whole changes and try to remember what was the previous state

**Comment by cdrage** [comment]
@benoitf I'll make sure it is separate commits now for any review updates.

Updated this PR with returning engineId

**Review by benoitf — APPROVED** [review]

**Review by deboer-tim — APPROVED** [review]

**Pull request merged**